exploration/selection-declaration.md

```diff
@@ -0,0 +1,296 @@
+Instead the selector's annotation replaces what came before.
+
+```
+.input {$num :integer}
+.match {$num :number minimumFractionDigits=1}
+* {{Formats {$num} like 1.0}}
+```
+
+**Pros**
+- Shorthand version works intuitively with minimal typing.
+
+**Cons**
+- Violates immutability that we've established everywhere else
+
+#### Allow "Declarative" Selectors with _Immutability_
+
+In this alternative, selectors are treated as declaration-selectors.
+That is, an annotation in a selector works like a `.input`.
+However, it is an error for the selector to try to modify a previous declaration
+(just as it is an error for a declaration to try to modify a previous declaration).
+This permits `.match` selectors to be a shorthand when no declarations exist.
+
+It is also an error for a selector to modify a previous selector.
+This implies that multiple selecton on the same operand is pointless.
+
+```
+.match {$num :integer}
+* {{Formats {$num} as integer}}
+
+.input {$num :integer}
+.match {$num :number maximumFractionDigits=0}
+* {{This message produces a Duplicate Declaration error}}
+
+.input {$num :integer} {$num :number}
+* * {{This message produces a Duplicate Declaration error}}
+```
+
+**Pros**
+- Shorthand version works intuitively with minimal typing
+- Preserves immutability
+- Produces an error when users inappropriately annotate some items
+
+**Cons**
+- Selectors can't provide additional selection-specific options
+  if the value has already been annotated
+- Doesn't allow multiple selection on the same operand, e.g.
+  ```
+  .input {$date :datetime skeleton=yMdjm}
+  .match {$date :datetime field=month} {$date :datetime field=dayOfWeek}
+  * * {{This message produces a Duplicate Declaration error
+        even though selection is separate from formatting.}}
+  ```
+
+#### Provide a `#`-like Feature
+
+(Copy-pasta adapted from @eemeli's proposal in #736)
+
+Make the `.match` expression also act as implicit declarations accessed by index position:
+
+```
+.match {$count :number}
+one {{You have {$0} apple}}
+* {{You have {$0} apples}}
+```
+
+Assigning values to `$0`, `$1`, ... would not conflict with any input values, 
+as numbers are invalid `name-start` characters.
+That's by design so that we encourage at least _some_ name for each variable; 
+here that's effectively provided by the `.match` expressions.
+
+ABNF would be modified:
+```diff
+-variable = "$" name
++variable = "$" (name / %x30-39)
+```
+
+...with accompanying spec language making numeric variables resolve to the `.match` selectors in placeholders, 
+and a data model error otherwise.
+
+**Pros**
+- More ergonomic for most `.input` cases
+- Enables representation of many messages without any declarations
+
+**Cons**
+- Confusing that the operand name can't be used in the pattern?
+  Removes some self-documentation from the pattern.
+- Requires the pattern to change if the selectors are modified.
+- Limits number of referenceable selectors to 10 (in the current form)
```

Comment:
I think one more alternative is missing, which is to do nothing: require a declaration when you want a declaration instead of introducing a third (maybe third and fourth?) declaration form, and maintain the invariant that the meaning of a variable name is the same regardless of where it appears.

Reply:
Will add